ibis/expr/scope.py

```diff
@@ -0,0 +1,176 @@
+""" Module for scope
+    The motivation of Scope is to cache data for calculated ops.
+
+    `scope` in Scope class is the main cache. It is a dictionary mapping
+    ibis node instances to concrete data, and the time context associate
+    with it (if any).
+
+    When there are no time contexts associate with the cached result, getting
+    and setting values in Scope would be as simple as get and set in a normal
+    dictonary. With time contexts, we need the following logic for getting
+    and setting items in scope:
+
+    Before setting the value op in scope we need to perform the following
+    check first:
+
+    Test if `op` is in `scope` yet
+    - No, then put `op` in `scope`, set 'timecontext' to be the current
+    `timecontext` (None if `timecontext` is not present), set 'value' to be
+    the actual data.
+    - Yes, then get the time context stored in `scope` for `op` as
+    `old_timecontext`, and compare it with current `timecontext`:
+    If current `timecontext` is a subset of `_timecontext`, that means we
+    already cached a larger range of data. Do nothing and we will trim data in
+    later execution process.
+    If current `timecontext` is a superset of `old_timecontext`, that means we
+    need to update cache. Set 'value' to be the current data and set
+    'timecontext' to be the current `timecontext` for `op`.
+    If current `timecontext` is neither a subset nor a superset of
+    `old_timcontext`, but they overlap, or not overlap at all (For example
+    when there is a window that looks forward, over a window that looks
+    back), in this case, we should not trust the data stored either because
+    the data stored in scope doesn't cover the current time context.
+    For simplicity, we update cache in this case, instead of merge data of
+    different time contexts.
+
+"""
+from collections import namedtuple
+from typing import Any, Dict, List, Optional
+
+from ibis.expr.operations import Node
+from ibis.expr.timecontext import TimeContextRelation, compare_timecontext
+from ibis.expr.typing import TimeContext
+
+ScopeItem = namedtuple('item', ['value', 'timecontext'])
+
+
+class Scope:
+    def __init__(self, items: Dict[str, ScopeItem] = None):
+        self.items = items or {}
```

Comment:
save this as `self._items`, then you can make a `.items()` method which return `iter(self._items)` and then Scope becomes more like a dictionary